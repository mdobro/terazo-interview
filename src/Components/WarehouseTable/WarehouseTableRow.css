.warehouseTableRow .actionsCell {
  white-space: nowrap;
}

.warehouseTableRow .editButton {
  margin-right: 2px;
}

.warehouseTableRow .tableInput {
  width: 100%;
}

.warehouseTableRow td {
  white-space: pre-line;
}
