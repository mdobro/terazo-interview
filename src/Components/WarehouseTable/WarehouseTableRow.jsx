import React, { useState } from "react";
import PropTypes from "prop-types";
import { Button } from "reactstrap";

import "./WarehouseTableRow.css";

const WarehouseTableRow = ({
  name,
  description,
  buildingName,
  streetLine1,
  streetLine2,
  city,
  stateProvince,
  zipPostalCode,
  country,
}) => {
  const [isEditing, setIsEditing] = useState(false);

  return (
    <tr className="warehouseTableRow">
      <td>{name}</td>
      <td>{description}</td>
      <td>{buildingName}</td>
      <td>{streetLine1}</td>
      <td>{streetLine2}</td>
      <td>{city}</td>
      <td>{stateProvince}</td>
      <td>{zipPostalCode}</td>
      <td>{country}</td>
      <td id="actionsCell">
        <Button color="success" size="sm" id="editButton">
          Edit
        </Button>
        <Button color="danger" size="sm">
          Delete
        </Button>
      </td>
    </tr>
  );
};

WarehouseTableRow.propTypes = {};

WarehouseTableRow.defaultProps = {};

export default WarehouseTableRow;
