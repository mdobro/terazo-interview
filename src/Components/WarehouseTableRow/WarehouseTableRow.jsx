import "./WarehouseTableRow.css";

import React, { useState } from "react";
import PropTypes from "prop-types";
import { Button, Input } from "reactstrap";
import { set } from "lodash";

const WarehouseTableRow = ({ warehouse, onRowChange, onRowDelete }) => {
  const [isEditing, setIsEditing] = useState(false);
  const [editedWarehouse, setEditedWarehouse] = useState(warehouse);

  const onSaveRow = () => {
    onRowChange(editedWarehouse);
    setIsEditing(false);
  };

  const onDeleteRow = () => {
    onRowDelete(warehouse);
  };

  const onInputChange = (e, fieldName) => {
    const { value } = e.target;

    setEditedWarehouse((prev) => {
      const newObj = { ...prev };
      set(newObj, fieldName, value);
      return newObj;
    });
  };

  const getCellContents = (data, fieldName) => {
    if (isEditing) {
      return (
        <Input
          data-testid="warehouseTableInput"
          type="textarea"
          className="tableInput"
          value={data ?? ""}
          onChange={(e) => onInputChange(e, fieldName)}
        />
      );
    }
    return data;
  };

  const {
    warehouseName,
    warehouseDescription,
    warehouseAddress: {
      buildingName,
      streetLine1,
      streetLine2,
      city,
      stateProvince,
      zipPostalCode,
      country,
    },
  } = editedWarehouse;

  return (
    <tr className="warehouseTableRow" data-testid="warehouseTableRow">
      <td>{getCellContents(warehouseName, "warehouseName")}</td>
      <td>{getCellContents(warehouseDescription, "warehouseDescription")}</td>
      <td>{getCellContents(buildingName, "warehouseAddress.buildingName")}</td>
      <td>{getCellContents(streetLine1, "warehouseAddress.streetLine1")}</td>
      <td>{getCellContents(streetLine2, "warehouseAddress.streetLine2")}</td>
      <td>{getCellContents(city, "warehouseAddress.city")}</td>
      <td>
        {getCellContents(stateProvince, "warehouseAddress.stateProvince")}
      </td>
      <td>
        {getCellContents(zipPostalCode, "warehouseAddress.zipPostalCode")}
      </td>
      <td>{getCellContents(country, "warehouseAddress.country")}</td>
      <td className="actionsCell">
        {isEditing ? (
          <Button
            color="success"
            size="sm"
            className="editButton"
            onClick={onSaveRow}
          >
            Save
          </Button>
        ) : (
          <>
            <Button
              color="success"
              size="sm"
              className="editButton"
              onClick={() => setIsEditing(true)}
            >
              Edit
            </Button>
            <Button color="danger" size="sm" onClick={onDeleteRow}>
              Delete
            </Button>
          </>
        )}
      </td>
    </tr>
  );
};

WarehouseTableRow.propTypes = {
  warehouse: PropTypes.shape({
    warehouseId: PropTypes.string,
    warehouseName: PropTypes.string,
    warehouseDescription: PropTypes.string,
    warehouseAddress: PropTypes.shape({
      buildingName: PropTypes.string,
      streetLine1: PropTypes.string,
      streetLine2: PropTypes.string,
      city: PropTypes.string,
      stateProvince: PropTypes.string,
      zipPostalCode: PropTypes.string,
      country: PropTypes.string,
    }).isRequired,
  }).isRequired,
  onRowChange: PropTypes.func,
  onRowDelete: PropTypes.func,
};

WarehouseTableRow.defaultProps = {
  onRowChange: undefined,
  onRowDelete: undefined,
};

export default WarehouseTableRow;
