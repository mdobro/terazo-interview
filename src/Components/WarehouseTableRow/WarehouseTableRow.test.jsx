/* eslint-disable no-restricted-globals */
import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import WarehouseTableRow from "./WarehouseTableRow";
import mockData from "../../mockData/warehouses.json";

const mockWarehouse = mockData[0];

const tbody = document.createElement("tbody");

test("renders", () => {
  const { asFragment } = render(
    <WarehouseTableRow warehouse={mockWarehouse} />,
    {
      container: document.body.appendChild(tbody),
    }
  );
  expect(asFragment()).toMatchSnapshot();
});

test("Edit", () => {
  const onRowChange = jest.fn();

  const { getByText, getAllByTestId } = render(
    <table>
      <tbody>
        <WarehouseTableRow
          warehouse={mockWarehouse}
          onRowChange={onRowChange}
        />
      </tbody>
    </table>
  );
  userEvent.click(getByText("Edit"));
  const inputs = getAllByTestId("warehouseTableInput");
  expect(inputs.length).toBe(9);

  userEvent.click(getByText("Save"));
  expect(onRowChange).toHaveBeenCalledWith(mockWarehouse);
});

test("Delete", async () => {
  const onRowDelete = jest.fn();

  const { getByTestId } = render(
    <table>
      <tbody>
        <WarehouseTableRow
          warehouse={mockWarehouse}
          onRowDelete={onRowDelete}
        />
      </tbody>
    </table>
  );
  userEvent.click(getByTestId("deleteButton"));

  userEvent.click(getByTestId("deleteModalButton"));
  expect(onRowDelete).toHaveBeenCalledWith(mockWarehouse);
});
