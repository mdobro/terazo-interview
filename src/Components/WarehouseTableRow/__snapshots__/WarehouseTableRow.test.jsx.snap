// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  Langosh-AdamsIn congue. Etiam justo. Etiam pretium iaculis justo.

In hac habitasse platea dictumst. Etiam faucibus cursus urna. Ut tellus.Daniel Inc509 Delladonna PlazaKibaraTanzania
  <button
    class="editButton btn btn-success btn-sm"
    type="button"
  >
    Edit
  </button>
  <button
    class="btn btn-danger btn-sm"
    data-testid="deleteButton"
    type="button"
  >
    Delete
  </button>
</DocumentFragment>
`;
